Ability to spawn/destroy lights and control light parameters

saihv opened the pull request:
### About
This PR adds the ability to spawn blueprints in the world along with static mesh actors. Additionally, it introduces a couple of sample blueprint wrappers for a Point light and a Spot light, and allows spawning or destroying them - which could be useful for dynamic changes in indoor environments. This method of control wraps the lights in a blueprint (see `Content/Blueprints/Lights`) that allows for spawning as well as easier, extensible control.

The main thing to note is that to be able to use this API, the light being targeted should be a blueprint. 

A new API function `simSetLightIntensity()` allows for changing the brightness of the lights. Example usage of the entire pipeline can be seen in `PythonClient/environment/light_control.py`

- [x] Ability to spawn blueprints
- [x] Brightness control
- [ ] Color control
- [ ] Unity placeholder API

### How Has This Been Tested?

Drag `Blueprints/Lights/PointLightBP.uasset` into any world (e.g. Blocks), and then run `PythonClient/environment/light_control.py`

### Video:


https://user-images.githubusercontent.com/2274262/131202787-95333e22-ef68-4dea-8dce-d875d76cd084.mp4

Commit:
Added ability to spawn blueprints, light control API

Commit:
Update docstring for simSpawnObject

rajat2004 commented on Unreal/Plugins/AirSim/Source/WorldSimApi.cpp:
```diff
@@ -153,6 +160,33 @@ AActor* WorldSimApi::createNewActor(const FActorSpawnParameters& spawn_params, c
+        const FString command = FString::Printf(TEXT("SetIntensity %f"), intensity);
+        FOutputDeviceNull ar;
+        actor->CallFunctionByNameWithArguments(*command, ar, NULL, true);
+        result = actor ? simmode_->getGlobalNedTransform().toGlobalNed(FTransform(actor->GetActorRotation(), actor->GetActorLocation()))
+                       : Pose::nanPose();
```
This line doesn't make sense, `result` is a bool, we're assigning a pose to it

saihv replied:
Whoops, this was a weird copy paste artifact I forgot to change.  Fixed

rajat2004 commented on Unreal/Plugins/AirSim/Source/WorldSimApi.cpp:
```diff
@@ -153,6 +160,33 @@ AActor* WorldSimApi::createNewActor(const FActorSpawnParameters& spawn_params, c
+                       : Pose::nanPose();
+    },
+                                             true);
+}
```
Missing `return result` here

saihv replied:
Fixed

rajat2004 commented on PythonClient/airsim/client.py:
```diff
@@ -158,6 +158,19 @@ def confirmConnection(self):
+
+        Args:
+            light_name (str): Name of light to change
+            intensity (float): New intensity value
```
Any range of this?

zimmy87 replied:
I tried playing around with the intensity property in LightActor instances placed in a scene. The UI sets a max value of 160.0 cd for the slider in the editor, but you can manually input a higher value with visible differences and no error. The source for ULightComponent::SetIntensity doesn't enforce any range, so any valid float value should work here. I'm fine with leaving the parameter description as is.

rajat2004 reviewed (commented):
I'm making almost the same changes in `spawnObject` in #3940 :laughing: (apart from the `is_blueprint`)
The API needs to be added in RpcLibClientBase and Unity files also

saihv committed:
Address review

saihv commented on PythonClient/environment/light_control.py:
```diff
@@ -0,0 +1,23 @@
+import airsim
+import time
+
+c = airsim.MultirotorClient()
```
Fixed

saihv commented on AirLib/include/api/WorldSimApiBase.hpp:
```diff
@@ -32,7 +32,7 @@ namespace airlib
         // ------ Level setting apis ----- //
         virtual bool loadLevel(const std::string& level_name) = 0;
-        virtual string spawnObject(string& object_name, const string& load_component, const Pose& pose, const Vector3r& scale, bool physics_enabled) = 0;
+        virtual string spawnObject(const string& object_name, const string& load_component, const Pose& pose, const Vector3r& scale, bool physics_enabled, bool is_blueprint) = 0;
```
Fixed

saihv commented on Unreal/Plugins/AirSim/Source/WorldSimApi.cpp:
```diff
@@ -82,61 +83,67 @@ bool WorldSimApi::destroyObject(const std::string& object_name)
+        std::vector<std::string> matching_names = UAirBlueprintLib::ListMatchingActors(simmode_, ".*" + final_object_name + ".*");
+        if (matching_names.size() > 0) {
+            size_t greatest_num{ 0 }, result{ 0 };
+            for (auto match : matching_names) {
```
Fixed

saihv commented on Unreal/Plugins/AirSim/Source/WorldSimApi.cpp:
```diff
@@ -82,61 +83,67 @@ bool WorldSimApi::destroyObject(const std::string& object_name)
+        // Ensure new non-matching name for the object
+        std::vector<std::string> matching_names = UAirBlueprintLib::ListMatchingActors(simmode_, ".*" + final_object_name + ".*");
+        if (matching_names.size() > 0) {
+            size_t greatest_num{ 0 }, result{ 0 };
```
Fixed

saihv commented on Unreal/Plugins/AirSim/Source/WorldSimApi.cpp:
```diff
@@ -82,61 +83,67 @@ bool WorldSimApi::destroyObject(const std::string& object_name)
+                std::string number_extension = match.substr(match.find_last_not_of("0123456789") + 1);
+                if (number_extension != "") {
+                    result = std::stoi(number_extension);
+                    greatest_num = greatest_num > result ? greatest_num : result;
```
Fixed

saihv commented on Unreal/Plugins/AirSim/Source/WorldSimApi.cpp:
```diff
@@ -153,6 +160,33 @@ AActor* WorldSimApi::createNewActor(const FActorSpawnParameters& spawn_params, c
 
+AActor* WorldSimApi::createNewBPActor(const FActorSpawnParameters& spawn_params, const FTransform& actor_transform, const Vector3r& scale, UBlueprint* blueprint)
+{
+    UClass* NewBP = (UClass*)blueprint->GeneratedClass;
```
Fixed

saihv commented on Unreal/Plugins/AirSim/Source/WorldSimApi.cpp:
```diff
@@ -153,6 +160,33 @@ AActor* WorldSimApi::createNewActor(const FActorSpawnParameters& spawn_params, c
+    UClass* NewBP = (UClass*)blueprint->GeneratedClass;
+    AActor* NewActor = simmode_->GetWorld()->SpawnActor<AActor>(NewBP, FVector::ZeroVector, FRotator::ZeroRotator, spawn_params);
+
+    if (NewActor) {
```
Fixed

saihv commented on Unreal/Plugins/AirSim/Source/WorldSimApi.cpp:
```diff
@@ -153,6 +160,33 @@ AActor* WorldSimApi::createNewActor(const FActorSpawnParameters& spawn_params, c
+{
+    bool result;
+    UAirBlueprintLib::RunCommandOnGameThread([this, &light_name, &intensity, &result]() {
+        AActor* actor = simmode_->scene_object_map.FindRef(FString(object_name.c_str()));
```
Fixed

rajat2004 commented on PythonClient/environment/light_control.py:
```diff
@@ -0,0 +1,23 @@
+
+# Change the light's intensity
+for i in range(20):
+    client.simSetLightIntensity("PointLight", i * 100)
```
Shouldn't we use the returned name from `simSpawnObject` instead of assuming that the object created got the same name?

zimmy87 replied:
fixed in latest iteration

rajat2004 commented on Unreal/Plugins/AirSim/Source/WorldSimApi.cpp:
```diff
@@ -153,6 +160,34 @@ AActor* WorldSimApi::createNewActor(const FActorSpawnParameters& spawn_params, c
+
+bool WorldSimApi::setLightIntensity(const std::string& light_name, float intensity)
+{
+    bool result;
```
Should be initialized to `false`

zimmy87 replied:
fixed in latest iteration

rajat2004 commented on Unreal/Plugins/AirSim/Source/WorldSimApi.cpp:
```diff
@@ -153,6 +160,34 @@ AActor* WorldSimApi::createNewActor(const FActorSpawnParameters& spawn_params, c
+        if (light_actor) {
+            const FString command = FString::Printf(TEXT("SetIntensity %f"), intensity);
+            FOutputDeviceNull ar;
+            result = light_actor->CallFunctionByNameWithArguments(*command, ar, NULL, true);
```
Maybe `nullptr` instead of `NULL`
Also, `result` should be set after this as well

zimmy87 replied:
fixed in latest iteration

zimmy87 reviewed (changes requested):
some style comments

zimmy87 committed:
Update WorldSimApi.cpp

address feedback from rajat2004

zimmy87 committed:
Update light_control.py

response to feedback from rajat2004

zimmy87 committed:
downgrade uassets

recreate PointLightBP.uasset & SpotLightBP.uasset so they're compatible with UE 4.25

zimmy87 committed:
Merge branch 'master' into PR/LightAPI

zimmy87 committed:
Update WorldSimApi.cpp

respond to feedback from zimmy87

zimmy87 committed:
fix Unity build

adding stub for new setLightIntensity API

zimmy87 committed:
apply style from clang-format

zimmy87 commented on Unreal/Plugins/AirSim/Source/WorldSimApi.cpp:
```diff
@@ -82,65 +84,70 @@ bool WorldSimApi::destroyObject(const std::string& object_name)
+std::string WorldSimApi::spawnObject(const std::string& object_name, const std::string& load_object, const WorldSimApi::Pose& pose, const WorldSimApi::Vector3r& scale, bool physics_enabled, bool is_blueprint)
 {
+    FString asset_name(load_object.c_str());
+    FAssetData* LoadAsset = simmode_->asset_map.Find(asset_name);
```
fixed in latest iteration

zimmy87 commented on Unreal/Plugins/AirSim/Source/WorldSimApi.cpp:
```diff
@@ -82,65 +84,70 @@ bool WorldSimApi::destroyObject(const std::string& object_name)
-            UAirBlueprintLib::setSimulatePhysics(NewActor, physics_enabled);
+        FActorSpawnParameters new_actor_spawn_params;
+        new_actor_spawn_params.Name = FName(final_object_name.c_str());
+        //new_actor_spawn_params.NameMode = FActorSpawnParameters::ESpawnActorNameMode::Required_ReturnNull;
```
fixed in latest iteration

zimmy87 commented on Unreal/Plugins/AirSim/Source/WorldSimApi.cpp:
```diff
@@ -153,6 +160,34 @@ AActor* WorldSimApi::createNewActor(const FActorSpawnParameters& spawn_params, c
+AActor* WorldSimApi::createNewBPActor(const FActorSpawnParameters& spawn_params, const FTransform& actor_transform, const Vector3r& scale, UBlueprint* blueprint)
+{
+    UClass* NewBP = static_cast<UClass*>(blueprint->GeneratedClass);
+    AActor* NewActor = simmode_->GetWorld()->SpawnActor<AActor>(NewBP, FVector::ZeroVector, FRotator::ZeroRotator, spawn_params);
```
fixed in latest iteration

LolkerXD commented:
😂
